source/AI.cpp

```diff
@@ -1235,9 +1240,11 @@ shared_ptr<Ship> AI::FindTarget(const Ship &ship) const
 				if(it->GetGovernment() != gov)
 				{
 					auto ptr = it->shared_from_this();
-					// Scan friendly ships that are as-yet unscanned by this ship's government.
-					if((!cargoScan || Has(gov, ptr, ShipEvent::SCAN_CARGO))
+					// Scan friendly ships that are as-yet unscanned by this ship's government,
+					// and are not in the process of being scanned.
+					if(((!cargoScan || Has(gov, ptr, ShipEvent::SCAN_CARGO))
 							&& (!outfitScan || Has(gov, ptr, ShipEvent::SCAN_OUTFITS)))
+							|| (Has(gov, target, ShipEvent::SCANNING) && !Has(ship, target, ShipEvent::SCANNING)))
```

Comment:
Looking at the code now, I realize this logic is used a lot (3 times, in fact) so I think it should be put in one place.
How about a small method; shouldScan, that would do these 3 lines, taking the ship, outfitscan, cargoscann and the target?

Dont worry about mustScanOutfit and mustScanCargo, they're useless bools and can be removed.

Reply:
In my original code the reason why shouldScan was separate was because the ship could carry illegal outfits or illegal cargo.  Not all ally ships have all types of scanning.  Some have only outfit scan and others have only cargo scan capabilities.  I think it makes sense to break out the two depending on context.

I guess shouldScan could be made a little smarter to return a bool depending on the outfits of the ship doing the scanning.  That's the only way I can think of to make shouldScan meaningful based on context.  Which class do you think shouldScan go into?  Could also be a namespace function.

Reply:
Yeah ofc we need a bool.
Well the thing is it's already using the Has() of AI so may as well not put it in another file than AI.cpp.
Now, since those are methods which need access to AI things, may as well make this an AI method as well?

Reply:
Yes, this expression should go in a private method of the AI class.
AI::Has() could become namespace methods if they were passed a (const) reference to the `AI::actions` map, but doing that seems excessive for this.

Reply:
I'm not sure how to act on this feedback.  Could you elaborate?